apps/next/src/memlab/playwright/test-builder.ts

```diff
@@ -0,0 +1,196 @@
+
+        // Setup phase (optional)
+        if (scenario.setup) {
+          await scenario.setup(page);
+        }
+
+        // Force GC and wait before baseline
+        await forceGC(page);
+        await waitForCleanup(page);
+
+        // BASELINE snapshot
+        await collector.takeSnapshot(page, 'baseline', 'before-action');
+
+        // TARGET: Perform the action
+        await scenario.action(page);
+        await waitForCleanup(page, scenario.actionWaitMs);
+
+        // TARGET snapshot
+        await collector.takeSnapshot(page, 'target', 'after-action');
+
+        // FINAL: Cleanup phase (optional)
+        if (scenario.cleanup) {
+          await scenario.cleanup(page);
+        }
+
+        // Force GC and wait before final snapshot
+        await forceGC(page);
+        await waitForCleanup(page, scenario.cleanupWaitMs);
+
+        // FINAL snapshot
+        await collector.takeSnapshot(page, 'final', 'after-cleanup');
+
+        // Analyze results
+        const result = await analyzeResult();
+        const maxLeaks = scenario.expectedLeaks ?? 0;
+
+        // Assert leak count is within threshold
+        expect(result.leakCount).toBeLessThanOrEqual(maxLeaks);
+      });
+    }
+  });
+}
+
+/**
+ * Create a stress test scenario that rapidly mounts/unmounts a component
+ *
+ * @param cycles - Number of mount/unmount cycles (default: 10)
+ * @param expectedLeaks - Maximum expected leaks (default: 5 for stress tests)
+ */
+export function createStressTestScenario(
+  cycles = 10,
+  expectedLeaks = 5,
+): TestScenario {
+  return {
+    name: 'stress test: rapid mount/unmount cycles',
+    description: `Rapidly toggles component ${cycles} times to detect accumulated leaks`,
+    action: async (page) => {
+      // Click the stress test button which handles the cycling
+      await page.click('[data-testid="stress-test"]');
+      // Wait for stress test to complete
+      await page.waitForSelector(
+        '[data-testid="stress-test"]:not([disabled])',
+        {
+          timeout: 30000,
+        },
+      );
+    },
+    expectedLeaks,
+    actionWaitMs: 1000,
+    cleanupWaitMs: 1000,
+  };
+}
+
+/**
+ * Create a basic mount/unmount test scenario
+ *
+ * @param toggleSelector - Selector for the toggle button
+ * @param contentSelector - Selector to verify component is mounted/unmounted
+ */
+export function createMountUnmountScenario(
+  toggleSelector: string,
+  contentSelector?: string,
+): TestScenario {
+  return {
+    name: 'mount/unmount cycle should not leak memory',
+    description: 'Basic visibility toggle to test component cleanup',
+    action: async (page) => {
+      // Toggle off
+      await page.click(toggleSelector);
+      if (contentSelector) {
+        await page
+          .waitForSelector(contentSelector, { state: 'hidden', timeout: 5000 })
+          // biome-ignore lint/suspicious/noEmptyBlockStatements: intentional swallow
+          .catch(() => {});
```

Comment:
curious: would a logger warning be helpful here and below instead of the empty catches?